### ISSUE TITLE
Detox support for multiple interacting devices?

### ISSUE THREAD

**Issue by sunweiyang** [wix/Detox#393]
I'm a developer for a RN video chat education app that connects pairs of users. Detox looks fantastic for testing single app instances, but I was not able to find anything about testing multiple app instances on more than one device interacting with each other. Is it possible to do this kind of testing with Detox on multiple simulators/emulators?

**Comment by LeoNatan** [comment]
Not possible at the moment without some work on your part. Theoretically you could run two different tests in two different shells, with two different configurations for each test. I assume this could work. 

That’s an interesting scenario. We plan to support concurrent testing on several devices, but that’s different.

**Comment by rotemmiz** [comment]
what @LeoNatan said. This currently out of the scope of the projects for us.

**Comment by ptmt** [comment]
I was able to achieve something similar for testing a chat app. Basically, it works as @LeoNatan described: two detox instances with different configurations and simple an http server for syncing between two instances (one may need to wait for another before performing an action) and sharing the context (such as sharing username just created).

![](https://media.giphy.com/media/3osBLd2ZB1nfvwV1Xq/giphy.gif)

Right now I did all basic tests such as sign in, search, exchanging messages. I'll continue with this approach with adding more and more tests, and if it works well, I'm planning to write a blog post or/and add an example repository.

**Comment by sunweiyang** [comment]
@ptmt This looks awesome! Please let us know when the post and/or example is live.

**Comment by ptmt** [comment]
There you go https://github.com/ptmt/inter-device-detox

It's an experiment and I'm really bad at writing notes for it. Anyway, it's a few lines of js, so you can get inspiration and do better. I'll continue experimenting with this since I found it really handy not just for tests, but for reproducing the environment from the scratch.

**Comment by rodrigoelp** [comment]
Hi @LeoNatan, do you know how far is the plan you mentioned above? (running concurrent devices with the same set of tests)

I am currently scripting a rake script (or fastlane) to run multiple instances of detox on a collection of iOS simulators (and possibly a collection of android emulators as well).

**Comment by rotemmiz** [comment]
It is supported on iOS from 7.4.0

**Comment by rodrigoelp** [comment]
I meant, running the test in variants of devices... iPhone 8, iPhone xs, iPhone SE, etc, at the same time. Is that documented anywhere?

I've been digging through the cli and I can see there is a way to provide the number of workers and maybe provide a device name with a given artefact so it doesn't need to recompile the code. But I've been unable to find a way to provide multiple simulators to test variations based on screen size or os variations.

**Comment by rotemmiz** [comment]
No, it's not supported, but you can just run it in two different commands, either one after the other, or by sending one one process to background with shell, something like `detox test -c ios.sim.iphonex& detox test -c ios.sim.iphone8`

**Comment by LeoNatan** [comment]
Probably a better idea to run both in background and then wait for both:
```
detox test -c ios.sim.iphonex&
detox test -c ios.sim.iphone8&
wait
```

**Comment by rotemmiz** [comment]
What he said

**Comment by manuquentin** [comment]
This should be fixed by #1144

**Comment by TammyTorres** [comment]
the fixed in which version is? Can we do detox test -o e2e/someconfig.json -p ios -a ci/the.ipa -devices "iPhone 8, iPhone XS, iPhone XS Max, iPhone SE" -c 'device configuration'. or have to specify each device separated? @manuquentin

**Comment by hsavit1** [comment]
> the fixed in which version is? Can we do
> ` detox test -o e2e/someconfig.json -p ios -a ci/the.ipa -devices "iPhone 8, iPhone XS, iPhone XS Max, iPhone SE" -c 'device configuration'.`
> 
> or have to specify each device separated? @manuquentin

Have you figured this out? Is this in the documentation somewhere?

**Comment by olexandr13** [comment]
+1

**Comment by d4vidi** [comment]
Reopened because we might want to support that one day.
